{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1209d39b-474d-4006-8c86-3ab865887516",
   "metadata": {},
   "outputs": [],
   "source": [
    "from gc import collect as gc_collect, disable as gc_disable\n",
    "from sys import maxsize, argv as sys_argv\n",
    "\n",
    "from time import time_ns\n",
    "from itertools import chain\n",
    "from numpy.lib.stride_tricks import sliding_window_view\n",
    "from operator import itemgetter\n",
    "from collections import defaultdict\n",
    "\n",
    "from numpy import frompyfunc, vectorize, uint8, uint32, array, concatenate, add, argmin, arange, column_stack, unique, split, empty, ix_, take, diff\n",
    "from bisect import bisect_right\n",
    "\n",
    "# Used to create kmer binary masks\n",
    "MAX_KMER_SIZE = 64\n",
    "\n",
    "# Mapping of special character used by suffix tables and similar structures\n",
    "MAPPING_DOLLAR = 0\n",
    "# Mapping of nucleotides\n",
    "MAPPING = dict(\n",
    "    C=1,\n",
    "    A=2,\n",
    "    T=3,\n",
    "    G=4,\n",
    ")\n",
    "# Allphabet\n",
    "ALPHABET = set(MAPPING.values())\n",
    "# Alphabet with included special character\n",
    "ALPHABET_DOLLAR = set([*ALPHABET, MAPPING_DOLLAR])\n",
    "\n",
    "# Pairs of alphabet symbols (a, b) such that a > b\n",
    "# Useful for O(a^2) iteration over alphabet\n",
    "ALPHABET_LT_PAIRS = [(a, b) for a in ALPHABET for b in ALPHABET if b < a]\n",
    "\n",
    "# Vectorized numpy function to map nucleotides from string\n",
    "MAPPING_FN = vectorize(MAPPING.get, otypes=[uint8])\n",
    "\n",
    "# Remove every nth kmer from the reference index\n",
    "REFERENCE_NTH_KMER_REMOVAL = 15 # was: 20\n",
    "# Load reference in chunks of that size\n",
    "REFERENCE_INDEX_CHUNK_SIZE = 800000\n",
    "\n",
    "# Length of k-mer used to generate (k,w)-minimizer indices\n",
    "KMER_LEN = 17 # was: 16\n",
    "# Length of window to generate (k,w)-minimizer indices\n",
    "WINDOW_LEN = 8 # was: 5\n",
    "\n",
    "# Match from k-mer index with be resized by kmer_len times this factor\n",
    "FACT_KMER_TO_RELATIVE_EXTENSION_LEN = 0.5\n",
    "\n",
    "# Miminal length of subsequent k-mers in LIS (longest increasing subsequence) that form a valid match\n",
    "MIN_LIS_EXTENSION_WINDOW_LEN = 3\n",
    "# Mimimum score for the extension window that is considered a match\n",
    "MIN_LIS_EXTENSION_WINDOW_SCORE = 0.1\n",
    "# Number of locally maximum scores for extension windows that we consider\n",
    "MAX_ACCEPTED_LIS_EXTENSION_WINDOWS_COUNT = 5\n",
    "# Max distance between starting k-mer of the match and the ending k-mer of match\n",
    "# Used when filtering the extended seed\n",
    "# Should be higher than FACT_TARGET_TO_QUERY_MAX_RELATIVE_LENGTH\n",
    "FACT_LIS_MAX_QUERY_DISTANCE = 1.3\n",
    "\n",
    "# Max difference between query and target to concude that it's a valid match\n",
    "#\n",
    "# if len(t_region) > len(q_region) * FACT_TARGET_TO_QUERY_MAX_RELATIVE_LENGTH:\n",
    "#   invalid_match()\n",
    "#\n",
    "FACT_TARGET_TO_QUERY_MAX_RELATIVE_LENGTH = 1.05\n",
    "# When using BWT aligner this is the maximum extra padding that we consider valid (as a fraction of len(query))\n",
    "FACT_BWT_QUERY_MAX_OFFSET = 0.04\n",
    "# When using BWT aligner this is the length of query part that we consider for fast matching (as a fraction of len(query))\n",
    "FACT_BWT_QUERY_FRAGMENT_SIZE = 0.1\n",
    "# When using BWT aligner this is the maximum number of error that we should encounter (to speed up searching)\n",
    "# The threshold is calculated as follows:\n",
    "#\n",
    "#    fragment_len = len(query) * FACT_BWT_QUERY_FRAGMENT_SIZE\n",
    "#    max_errors = fragment_len * FACT_BWT_FRAGMENT_REL_ERRORS_THRESHOLD\n",
    "#\n",
    "FACT_BWT_FRAGMENT_REL_ERRORS_THRESHOLD = 0.10\n",
    "\n",
    "# When using DP aligner (more accurate than BWT, but also slower)\n",
    "# we consider pairs (kmer_len, kmer_skip)\n",
    "# kmer_len is the length of k-mer we consider\n",
    "# kmer_skip means we skip every n-th kmer \n",
    "# Those values can be significantlly lower than global KMER_SIZE, because we run DP aligner only in specific situations\n",
    "# If no match is found we use the next configuration untill we find anything\n",
    "DP_K_STEP_SEQUENCE = [(15, 11), (8, 5)]\n",
    "# Length of the query suffix we use for the DP aligner as a fraction of query length\n",
    "FACT_DP_QUERY_SUFFIX_REL_LEN = 0.4\n",
    "# For DP aligner we set maxiumum exit distance \n",
    "# This value is fraction of the length of the query\n",
    "FACT_DP_QUERY_REL_MAX_E_DISTANCE = 0.1111\n",
    "\n",
    "# Cost when gap is opened (DP aligner uses only COST_GAP_EXTEND for gaps)\n",
    "COST_GAP_OPEN = 3\n",
    "# Cost when gap is extended (DP aligner uses only COST_GAP_EXTEND for gaps)\n",
    "COST_GAP_EXTEND = 1\n",
    "# Cost of mismatch\n",
    "COST_MISMATCH = 1\n",
    "# Cost of match\n",
    "COST_MATCH = 0\n",
    "\n",
    "# Globals used to cache values and speed up calculations (no need to pass around many references in call frames)\n",
    "# C[a] := number of lexicographically smaller letters than a in bw/reference\n",
    "_global_bwt_c = {}\n",
    "# O is a dictionary with keys $,A,C,G,T, and values are arrays of counts\n",
    "_global_bwt_o = {}\n",
    "# D[i] := lower bound on number of differences in substring s[1:i]\n",
    "_global_bwt_d = []\n",
    "# mask[k] := mask used to calculate hash for k-mer of length k\n",
    "_global_masks = None\n",
    "\n",
    "# Makes sure 0 <= pos <= len\n",
    "def normalize_pos(pos, len):\n",
    "    return min(max(pos, 0), len)\n",
    "\n",
    "# Generate hash mask for given kmer length\n",
    "# The mask will be just 2*k last bits on for given k-mer len: k\n",
    "# e.g mask(3) = b...000000111111\n",
    "def generate_mask(\n",
    "    kmer_len: int,\n",
    ") -> int:\n",
    "    global _global_masks\n",
    "    if not _global_masks:\n",
    "        _global_masks = dict()\n",
    "        ret = 3\n",
    "        for i in range(MAX_KMER_SIZE+1):\n",
    "            ret = (ret << 2) | 3\n",
    "            _global_masks[i] = ret\n",
    "    return _global_masks[kmer_len]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "acacef1b-07d1-404f-b965-165219268c32",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">SRR27257243.1 ST-E00314:339:HCCGHCCXY:3:1101:20517:1907/1\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import gzip\n",
    "for line in gzip.open('../data_classif/SRR27257243_1.fasta.gz', 'rt'):\n",
    "    print(line)\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "dd76e3f8-69e6-4b91-9604-b2b100645dae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[(0, 'A'), (1, 'B')]\n"
     ]
    }
   ],
   "source": [
    "print(list(enumerate([\"A\", \"B\"])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18a05da1-5a85-4b4c-a573-374c50eee7ee",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
